connections/stream-migration.md

```diff
@@ -0,0 +1,217 @@
+
+== Stream Migration ==
+
+note over Initiator, Responder: Migrate <b>Stream A</b> to <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream A:</b> Start stream migration, this stream is <b>id=A</b>
+
+Initiator -> Responder: Open new stream on <b>Connection 2</b>. Call this <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream B:</b> Migrate stream with <b>id=A</b> to this stream
+Initiator <- Responder: <b>Stream B:</b> Ack Migrate
+
+note over Initiator
+    Close <b>stream A</b> for writing.
+    Will only write to <b>stream B</b> from now on.
+end note
+
+
+note over Initiator
+    We have already seen the ""EOF"" on
+    <b>Stream A</b> from <i>Responder</i>
+    So we continue reading on <b>stream B</b>
+end note
+
+Initiator -> Responder: <b>Stream A:</b> ""EOF""
+
+note over Responder
+    Treat ""EOF"" on <b>stream A</b> as a signal to close <b>stream A</b> for
+    writing and continue writing on <b>stream B</b>. However stream A was
+    already closed (before migration), so we close <b>stream B</b> as well here.
+end note
+Initiator <- Responder: <b>Stream B:</b> ""EOF""
+
+
+note over Initiator, Responder: Stream A is now migrated to Stream B
+@enduml
+```
+To generate:
+```bash
+plantuml stream-migration.md -o stream-migration -tsvg
+```
+</details>
+
+The reverse case where the Initiator's stream is closed for writing is the same
+as above, but mirrored.
+
+## Picking the best connection
+
+Moving streams from one connection to another involves picking which connection
+we should move the streams to. Here are some recommended heuristics the
+initiator may use in determining which connection is best.
+
+1. If we have both relayed and direct connections, keep the direct connections
+   and drop the relay connections.
+2. Check for simultaneous connect: If we have both inbound and outbound
+   connections, keep the ones initiated by the peer with the lowest peer ID. Open
+   Question: Some protocols behave differently depending on whether they are the
+   dialer or listener. Can we really consolidate these?
+3. Prefer the connection with the most streams.
+4. Break ties in the remaining connections by selecting the newest conn, to
+   match the swarm's behavior in best connection selection.
+
+Note that it's not required that all implementations (and all versions) follow
+the same heuristics since the initiator is driving the migration and specifies
+where to migrate to.
+
+## Appendix
+
+[Specs Issue](https://github.com/libp2p/specs/issues/328)
+
+### Related Issues:
+
+- <https://github.com/libp2p/go-libp2p/issues/634>
+
+## Open Questions
+
+Some questions that will probably be resolved when a PoC is implemented.
+
+- How many streams are there total? 2 or 3? There is the old stream and new
+  stream, but is there a migration protocol specific stream? Or can the
+  migration be started from the old stream? If yes, how do we prevent
+  application data from looking like a migration without framing?
```

Comment:
I interpret your first diagram as saying that the migration is initiated on stream B by saying “migrate stream A to this stream”. Is this not what you meant?

Reply:
Yeah, that's my current thinking. but I think there's a difference in how stebalien presented the protocol in #328. Specifically step 3 in the first part. Maybe I'm missing something but I don't think you need that part (updated the diagram to reflect that). I think you only need the new stream that knows what the stream id is of the old stream.